### PR TITLE
 Wrb proxy

### LINKED ISSUE #57: WB smart proxy
The goal is to have a Witnet Bridge Proxy that complies with the WB interface. This proxy should maintain a state that:

- Stores, for each Witnet Bridge Controller, which is the latest request ID that has been issued.
- Define the necessary functions to upgrade the Witnet Bridge Controller

### PR DESCRIPTION
**Pull request by clbartoli**
Added the WRB Proxy contact with the methods:
- postDataRequest
- upgradeDataRequest
- readResult
- upgradeWitnetRequestsBoard
- getController

Added tests for WRB Proxy

Added changes in UsingWintet and UsingWitnetBytes, now they call the WRB Proxy
Added changes in the tests of both  UsingWintet and UsingWitnetBytes

This PR closes #57

### CODE REVIEW

File: contracts/WitnetRequestsBoardInterface.sol [hunk 1]
```diff
@@ -1,15 +1,15 @@
 pragma solidity ^0.5.0;
 /**
- * @title WBI Interface
- * @notice Interface of a WBI
- * It defines how to interact with the WBI in order to support:
+ * @title Witnet Requests Board Interface
+ * @notice Interface of a Witnet Request Board
+ * It defines how to interact with theRB in order to support:
```
Comment (girazoki) [inline contracts/WitnetRequestsBoardInterface.sol]:
```suggestion
 * It defines how to interact with the WRB in order to support:
```

File: contracts/WitnetRequestsBoardProxy.sol [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    require(_newAddress != witnetRequestsBoardAddress, "The provided Witnet Requests Board instance address is already in use");
+    _;
+  }
+
+  constructor(address _witnetRequestsBoardAddress) public {
+    witnetRequestsBoardAddress = _witnetRequestsBoardAddress;
+    witnetRequestsBoardInstance = WitnetRequestsBoardInterface(_witnetRequestsBoardAddress);
+    // Set the first possition of the lastIdsControllers to zero
+    lastIdsControllers = [0];
+  }
```
Comment (girazoki) [inline contracts/WitnetRequestsBoardProxy.sol]: Push the _witnetRequestsBoardAddress to the mapping/strcuture too as initialization

File: contracts/WitnetRequestsBoardProxy.sol [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  /// @dev Increments the rewards of a data request by adding more value to it. The new request reward will be increased by msg.value minus the difference between the former tally reward and the new tally reward.
+  /// @param _id The unique identifier of the data request.
+  /// @param _tallyReward The new tally reward. Needs to be equal or greater than the former tally reward.
+  function upgradeDataRequest(uint256 _id, uint256 _tallyReward)
+    external
+    payable
+  {
+    return witnetRequestsBoardInstance.upgradeDataRequest(_id, _tallyReward);
```
Comment (girazoki) [inline contracts/WitnetRequestsBoardProxy.sol]: _id-offset

File: contracts/WitnetRequestsBoardProxy.sol [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    uint256 offSetWrb;
+    address wrbAddress;
+    // If the id is bigger than the lastId of the previous Controller, read the result in the current Controller
+    if (_id > lastIdsControllers[n - 1]) {
+      return (witnetRequestsBoardAddress, lastIdsControllers[n - 1]);
+    } else {
+      // Else check the first lastId so that is bigger
+      for (uint i = 0; i <= n - 1; i++) {
+        if (_id > lastIdsControllers[n - 1 - i]) {
+          WitnetRequestsBoardInterface wrbWithResult;
```
Comment (girazoki) [inline contracts/WitnetRequestsBoardProxy.sol]: Same as in block relay

File: contracts/UsingWitnetBytes.sol [hunk 5]
```diff
@@ -38,9 +38,9 @@ contract UsingWitnetBytes {
   * @param _id The id of a request that has been previously sent to the WitnetRequestsBoard.
   * @return A boolean telling if the request has been already accepted or not. `false` do not mean rejection, though.
   */
-  function witnetCheckRequestAccepted(uint256 _id) internal view returns(bool) {
+  function witnetCheckRequestAccepted(uint256 _id) internal returns(bool) {
```
Comment (mariocao) [inline contracts/UsingWitnetBytes.sol]: It is possible to keep the `view` modifier?

If I am not mistaken, this function uses `readDrHash` and then `getController`. 
I think, if we set `getController` to view it seems we could keep the `view` modifier.

File: contracts/WitnetRequestsBoard.sol [hunk 6]
```diff
@@ -35,6 +35,8 @@ contract WitnetRequestsBoard is WitnetRequestsBoardInterface {
 
   ActiveBridgeSetLib.ActiveBridgeSet abs;
 
+  address witnet;
```
Comment (girazoki) [inline contracts/WitnetRequestsBoard.sol]: comment on what this is?

File: contracts/WitnetRequestsBoardProxy.sol [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+/**
+ * @title Block Relay Proxy
+ * @notice Contract to act as a proxy between the Witnet Bridge Interface and the block relay
+ * @dev More information can be found here
+ * DISCLAIMER: this is a work in progress, meaning the contract could be voulnerable to attacks
+ * @author Witnet Foundation
+ */
+contract WitnetRequestsBoardProxy {
+
+  address public witnetRequestsBoardAddress;
```
Comment (girazoki) [inline contracts/WitnetRequestsBoardProxy.sol]: Comments on the state variables

File: contracts/WitnetRequestsBoardProxy.sol [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+  /// @notice Upgrades the Witnet Requests Board if the current one is upgradeable
+  /// @param _newAddress address of the new block relay to upgrade
+  function upgradeWitnetRequestsBoard(address _newAddress) public notIdentical(_newAddress) {
+    // Require the WRB is upgradable
+    require(witnetRequestsBoardInstance.isUpgradable(msg.sender), "The upgrade has been rejected by the current implementation");
+    // Set the offSet for the next WRB
+    uint256 n = controllers.length;
+    // Map the currentLastId to the corresponding witnetRequestsBoardAddress and add it to controllers
+    // controllers.push(ControllerInfo({controllerAddress: _newAddress, lastId: currentLastId, offset: controllers[n - 1].lastId}));
```
Comment (girazoki) [inline contracts/WitnetRequestsBoardProxy.sol]: remove this line

File: contracts/WitnetRequestsBoardProxy.sol [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+  address public witnetRequestsBoardAddress;
+  WitnetRequestsBoardInterface witnetRequestsBoardInstance;
+  // Last id of the WRB controller
+  uint256 currentLastId;
+
+  // Struct if the information of each controller
+  struct ControllerInfo {
+    address controllerAddress;
+    uint256 lastId;
```
Comment (girazoki) [inline contracts/WitnetRequestsBoardProxy.sol]: It seems to be offset and lastId are identical, so try to remove one

File: test/helpers/WrbProxyTestHelper.sol [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  function getControllerAddress(uint256 _id) public returns(address) {
+    address wrb;
+    uint256 offset;
+    (wrb, offset) = getController(_id);
+    return wrb;
+  }
+
+//   function _witnetUpgradeDataRequest(uint256 _id, uint256 _tallyReward) public payable {
```
Comment (girazoki) [inline test/helpers/WrbProxyTestHelper.sol]: remove commented functions

**Pull request merged**